### ISSUE TITLE
`(svd

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#69531]
## To Reproduce

1.
1.
1.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by Lezcano** [comment]
At the moment, this function is autodifferentiated. This is wrong. This function should have its own backward method.

Note that, for this function, we have that `A = USVh + E`, so a correction term might be needed in the usual SVD formula. Besides that, I think that we would just need to modify a little bit `svd_backward` to not use the assumption that `U.shape = (m, k), Vh.shape = (k,n)` with `k = min(m, n)`, but simply `k <= min(m, n)`. This second part should not be too difficult.

**Comment by nikitaved** [comment]
The svd_backward method handles any shapes. I would go with a low-rank SVD implementation that handles any shape, but in it's backward takes care of the input's rank to pass write arguments to svd_backward.

**Comment by Lezcano** [comment]
It assumes that you give it `k = min(m, n)` non-zero singular values (i.e. full rank). Here we would need `k <= min(m, n)`, so we cannot use `svd_backward` as is.

In any case, it's not clear why doing autodiff through the `svd_lowrank` function makes any sense.

**Comment by nikitaved** [comment]
After checking the code it is not obvious to me why it would not handle the case with `k < min(m, n)`.

> In any case, it's not clear why doing autodiff through the `svd_lowrank` function makes any sense.
You mean in general? If so, I disagree. The backward is still well-defined for inputs with non-repeating non-zero singular values.

**Comment by Lezcano** [comment]
No, I mean autodiff. What I'm saying is that the function should have a custom backward. You cannot run autodiff through an algorithm that approximates something without proving that it is indeed correct to do so (i.e. the algorithm approximates the function to first order, not only to 0-th order).

**Comment by nikitaved** [comment]
With that I most certainly agree.
